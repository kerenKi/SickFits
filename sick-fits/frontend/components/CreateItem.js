import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';
import Form from './styles/Form';
import formatMoney from '../lib/formatMoney';
import Error from './ErrorMessage';

//query for the mutation
const CREATE_ITEM_MUTATION = gql`
mutation CREATE_ITEM_MUTATION (
    $title: String!,
    $description: String!,
    $image: String,
    $largeImage: String,
    $price: Int!
) {
  createItem(
    title: $title,
    description: $description,
    image: $image,
    largeImage: $largeImage,
    price: $price
  ) {
    id
  }
}
`;

class CreateItem extends Component {
  state = {
    title: 't-shirt',
    description: 'mmmmm',
    image: 'dog.jpg',
    largeImage: 'doggg.jpg',
    price: 100,
  }
  handleChange = (event) => {
    const { name, type, value } = event.target
    const val = type === 'number' ? parseFloat(value) : value
    this.setState({
      ...this.state,
      [name]: val
    })
  }
  render() {
    return <Mutation mutation={ CREATE_ITEM_MUTATION } variables={this.state}> 
      { (createItem, {loading, error}) => (
        <Form onSubmit={ async (e) => {
          //stop the form from reloading
          e.preventDefault()
          //call the mutation
          const res = await createItem()
          console.log(res)
          //change them to the single item page
          Router.push({
            pathname:'/item',
            query: {id: res.data.createItem.id }
          })

        }}>
          <h2>Sell an Item</h2>
          <Error error={error}/>
          <fieldset disabled={loading} aria-busy={loading}>
            <label htmlFor="title">
              Title
              <input 
                type="text" 
                id="title" 
                name="title" 
                placeholder="Title" 
                required
                value={this.state.title}
                onChange={this.handleChange}
                />
            </label>

            <label htmlFor="price">
              Price
              <input 
                type="number" 
                id="price" 
                name="price" 
                placeholder="Price" 
                required
                value={this.state.price}
                onChange={this.handleChange}
                />
            </label>

            <label htmlFor="description">
              Description
              <textarea 
                type="text" 
                id="description" 
                name="description" 
                placeholder="Enter A Description" 
                required
                value={this.state.description}
                onChange={this.handleChange}
                />
            </label>
            <button type="submit"> Submit</button>
          </fieldset>
        </Form>
    )}
    </Mutation> 
  }
}

export default CreateItem;
export { CREATE_ITEM_MUTATION };