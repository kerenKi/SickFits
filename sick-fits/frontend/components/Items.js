import React, { Component } from 'react';
// Imports for making a query
import { Query } from 'react-apollo';
import gql from 'graphql-tag';

//It's best practice to put all queries in caps
//queries should be named so it is easy to call them the same as their variable
const ALL_ITEMS_QUERY = gql`
  query ALL_ITEMS_QUERY {
    items {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;

//we use the query through render props - we have a Query component that takes our ALL_ITEMS_QUERY as a prop
//The only child of a query component must be a function
class Items extends Component {
  render() {
    return (
      <div>
        <p>Items!</p>
        <Query query={ALL_ITEMS_QUERY}>
          {(payload) => {
            console.log(payload)
            return <p>Hi! Query</p>
          }}
        </Query>
      </div>
    );
  }
}

export default Items;