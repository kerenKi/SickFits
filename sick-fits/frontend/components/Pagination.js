import React from 'react';
import PaginationStyles from './styles/PaginationStyles';

const Pagination = (props) => (
    <PaginationStyles>
      <p>here is the pagination</p>
    </PaginationStyles>
)

export default Pagination;