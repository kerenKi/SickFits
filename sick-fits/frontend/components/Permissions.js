import React, { Component } from 'react';
import { Query } from 'react-apollo';
import Error from './ErrorMessage';
import gql from 'graphql-tag';
import Table from './styles/Table';
import SickButton from './styles/SickButton';
import PropTypes from 'prop-types';


const possiblePermissions = [
  'ADMIN',
  'USER',
  'ITEMCREATE',
  'ITEMUPDATE',
  'ITEMDELETE',
  'PERMISSIONUPDATE',
]

const ALL_USERS_QUERY = gql`
  query ALL_USERS_QUERY {
    users {
      id
      name
      email
      permissions
    }
  }
`;

const Permissions = props => {
  return (
    <Query query={ ALL_USERS_QUERY }>
      { payload => {
        const { data, loading, error } = payload
        console.log(data)

        return (
          <div>
            <Error error={error}/>
            <div>
              <h2>Manage Permissions</h2>
              <Table>
                <thead>
                  <tr>
                    <th>Name</th>
                    <th>Email</th>
                    {possiblePermissions.map(
                      permission => <th key={permission}>{permission}</th>
                    )}
                    <th>👇🏻</th>
                  </tr>
                </thead>
                <tbody>
                  {data.users.map(user => <UserPermissions key={user.id} user={user}/>)}
                </tbody>
              </Table>
            </div>
          </div>
        )
      }}
    </Query>
  )
}

class UserPermissions extends React.Component {
  static propTypes = {
    user: PropTypes.shape({
      name: PropTypes.string,
      email: PropTypes.string,
      id: PropTypes.string,
      permissions: PropTypes.array,
    }).isRequired
  }

  state= {
    permissions: this.props.user.permissions,
  }

  handlePermissionChange = (event) => {
    const checkbox = event.target
    let updatedPermissions = [...this.state.permissions]
    if(checkbox.checked) {
      //add in the value to permissions list
      updatedPermissions.push(checkbox.value)
    } else {
      //remove the permission from the list
      updatedPermissions = updatedPermissions.filter( permission => permission !== checkbox.value)
    }

    this.setState({
      permissions: updatedPermissions,
    })
  }

  render() {
    const user = this.props.user
    return (
      <tr>
        <td>{user.name}</td>
        <td>{user.email}</td>
        {possiblePermissions.map( permission => (
          <td key={permission}>
            <label htmlFor={`${user.id}-permission-${permission}`}>
              <input
                id={`${user.id}-permission-${permission}`}
                type="checkbox"
                checked={this.state.permissions.includes(permission)}
                value={permission}
                onChange={this.handlePermissionChange}
              />
            </label>
          </td>
          ))}
          <td>
            <SickButton> Update</SickButton>
          </td>
      </tr>
    )
  }

}

export default Permissions;

