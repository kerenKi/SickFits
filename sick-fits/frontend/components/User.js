import React, { Component } from 'react';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import PropTypes from 'prop-types';
import Error from './ErrorMessage';
import styled from 'styled-components';
import Head from 'next/head'


const CURRENT_USER_QUERY = gql`
  query CURRENT_USER_QUERY {
    me {
      id
      email
      name
      cart {
        id
        quantity
        item {
          id
          title
          image
          price
          description
        }
      }
    }
  }
`;

const User = (props) => (
  <Query  {...props} query={CURRENT_USER_QUERY}>
    {payload => props.children(payload) }
  </Query>
)

User.propTypes = {
  children: PropTypes.func.isRequired,
}
export default User;
export { CURRENT_USER_QUERY };