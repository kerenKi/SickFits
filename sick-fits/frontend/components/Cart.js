import React, { Component } from "react";
import { Query, Mutation } from "react-apollo";
import gql from "graphql-tag";
import { adopt } from "react-adopt";
import User from "./User";
import CartStyles from "./styles/CartStyles";
import Supreme from "./styles/Supreme";
import CloseButton from "./styles/CloseButton";
import SickButton from "./styles/CartStyles";
import CartItem from "./CartItem";
import calcTotalPrice from "../lib/calcTotalPrice";
import formatMoney from "../lib/formatMoney";

const LOCAL_STATE_QUERY = gql`
  query LOCAL_STATE_QUERY {
    cartOpen @client
  }
`;

const TOGGLE_CART_MUTATION = gql`
  mutation TOGGLE_CART_MUTATION {
    toggleCart @client
  }
`;
// using react adopt to compose all the different components and avoid nesting
const Composed = adopt({
  //To avoid the error: index.js:2178 Warning: Failed prop type: The prop `children` is marked as required in `User`, but its value is `undefined`.
  //in User (at Cart.js:28)
  //we make the user a function, destructure the render method and then pass it as a child to the User component
  user: ({ render }) => <User>{render}</User>,
  toggleCart: ({ render }) => (
    <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>
  ),
  localState: ({ render }) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>
});

const Cart = props => {
  return (
    <Composed>
      {payload => {
        const { user, toggleCart, localState } = payload;
        const me = user.data.me;
        if (!me) {
          return null;
        }
        return (
          <CartStyles open={localState.data.cartOpen}>
            <header>
              <CloseButton title="close" onClick={toggleCart}>
                &times;
              </CloseButton>
              <Supreme> {me.name}'s cart </Supreme>
              <p>
                You have {me.cart.length} item{me.cart.length == 1 ? "" : "s"}{" "}
                in your cart
              </p>
            </header>
            <ul>
              {me.cart.map(cartItem => (
                <CartItem key={cartItem.id} cartItem={cartItem} />
              ))}
            </ul>
            <footer>
              <p>{formatMoney(calcTotalPrice(me.cart))}</p>
              <SickButton>Checkout</SickButton>
            </footer>
          </CartStyles>
        );
      }}
    </Composed>
  );
};

export default Cart;
export { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION };
