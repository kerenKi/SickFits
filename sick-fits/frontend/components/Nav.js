import React from 'react';
import Link from 'next/link';
import NavStyles from './styles/NavStyles'
import User from './User';

function Nav(props) {
  return (
      <User>
        { (payload) => {
          const { me } = payload.data
          return (
          <NavStyles>
            <Link href="/items">
              <a>Shop</a>
            </Link>
            { me && (
              <>
                <Link href="/sell">
                  <a>Sell</a>
                </Link>
                
                <Link href="/orders">
                  <a>Orders</a>
                </Link>

                <Link href="/me">
                  <a>Account</a>
                </Link>
              </>
            )}
            
            {!me && <Link href="/signup">
              <a>Signin</a>
            </Link>}
            
          </NavStyles>
          )}
        }
      </User>
      
  );
}

export default Nav;