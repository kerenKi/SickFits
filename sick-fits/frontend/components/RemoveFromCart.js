import React, { Component } from 'react';
import { CURRENT_USER_QUERY } from './User';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const REMOVE_FROM_CART_MUTATION = gql`
  mutation REMOVE_FROM_CART_MUTATION ($id: ID!) {
    removeFromCart (id: $id) {
      id
    }
  }
`;

const BigButton = styled.button`
  font-size: 3rem;
  background: none;
  border: 0;
  &:hover {
    color: ${ props => props.theme.red };
    cursor: pointer;
  }
`;

class RemoveFromCart extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
  }

  // update gets called when the response coe back from the server after a nutation
  update =  (cache, payload) => {
    //read the cache
    const data = cache.readQuery({ query : CURRENT_USER_QUERY })
    console.log(data)
    //remove item from cart
    const itemToRemoveId = payload.data.removeFromCart.id
    data.me.cart = data.me.cart.filter( item => item.id !== itemToRemoveId)
    //write back to cache
    cache.writeQuery({ query : CURRENT_USER_QUERY, data: data })
  }

  render() {
    return (
      <Mutation
        mutation={ REMOVE_FROM_CART_MUTATION }
        variables={{ id: this.props.id }}
        update={this.update}
        optimisticResponse={{
          __typename: 'Mutation',
          removeFromCart: {
            __typename: 'CartItem',
            id: this.props.id,
          }
        }}
        >
        {(removeFromCart, { loading }) => (
          <BigButton
            title="Delete Item"
            onClick={()=> {
              removeFromCart().catch(err => alert(err.message))
            }}
            disabled={loading}
          >&times;</BigButton>
        )}
      </Mutation>
    );
  }
}

export default RemoveFromCart;