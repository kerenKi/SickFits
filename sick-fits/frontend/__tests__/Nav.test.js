import { mount } from "enzyme";
import wait from "waait";
import Nav from "../components/Nav";
import { CURRENT_USER_QUERY } from "../components/User";
import { MockedProvider } from "react-apollo/test-utils";
import { fakeUser, fakeCartItem } from "../lib/testUtils";
import toJSON from 'enzyme-to-json'


const notSignedInMocks = [
  {
    request: {
      query: CURRENT_USER_QUERY
    },
    result: {
      data: { me: null }
    }
  }
];

const SignedInMocks = [
  {
    request: {
      query: CURRENT_USER_QUERY
    },
    result: {
      data: { me: fakeUser() }
    }
  }
];

const SignedInMocksWithCartItems = [
  {
    request: {
      query: CURRENT_USER_QUERY
    },
    result: {
      data: { me: {
        ...fakeUser(),
        cart: [fakeCartItem(), fakeCartItem(), fakeCartItem(), fakeCartItem()]
      } }
    }
  }
];

describe('<Nav/>', () => {
  it('should render signin link if user is not logged in', async () => {
    const wrapper = mount(
      <MockedProvider mocks={notSignedInMocks}>
        <Nav/>
      </MockedProvider>
    )
    await wait()
    wrapper.update()
    // console.log('wrapper :', wrapper.debug());
    const link = wrapper.find('a')
    // console.log('link', link.debug())
    //check if there are 2 links when not signed in
    expect(wrapper.find('a')).toHaveLength(2)

    const nav = wrapper.find('ul[data-test="nav"]')
    // console.log('nav :', nav.debug()); 
    //screenshot test
    expect(toJSON(nav)).toMatchSnapshot()
  })

  it('renders full nav links when user is signed in', async () => {
    const wrapper = mount(
      <MockedProvider mocks={SignedInMocks}>
        <Nav/>
      </MockedProvider>
    )
    await wait()
    wrapper.update()
    // console.log('wrapper :', wrapper.debug());
    const nav = wrapper.find('ul[data-test="nav"]')
    // expect(toJSON(nav)).toMatchSnapshot()
    expect(nav.children().length).toBe(6)
    expect(nav.text()).toContain('Sign Out')

  })

  it('renders the amount of items in the cart', async () => {
    const wrapper = mount(
      <MockedProvider mocks={SignedInMocksWithCartItems}>
        <Nav/>
      </MockedProvider>
    )
    await wait()
    wrapper.update()
    // console.log('wrapper :', wrapper.debug());
    const nav = wrapper.find('[data-test="nav"]')
    const count = nav.find('div.count')
    // console.log('count :', count.text());
    expect(toJSON(count)).toMatchSnapshot()
  })
})