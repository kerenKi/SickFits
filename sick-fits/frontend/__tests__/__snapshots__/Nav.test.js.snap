// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> renders the amount of items in the cart 1`] = `
<div
  className="count CartCount__Dot-xxvp4g-1 fJsVOg"
>
  12
</div>
`;

exports[`<Nav/> should render signin link if user is not logged in 1`] = `
<ul
  className="NavStyles-sc-11c0d2g-0 iTUtFG"
  data-test="nav"
>
  <Link
    href="/items"
  >
    <a
      href="/items"
      onClick={[Function]}
    >
      Shop
    </a>
  </Link>
  <Link
    href="/signup"
  >
    <a
      href="/signup"
      onClick={[Function]}
    >
      Signin
    </a>
  </Link>
</ul>
`;
